rust/noosphere-storage/src/implementation/native.rs

```diff
@@ -99,3 +98,57 @@ impl Drop for NativeStorage {
         let _ = self.db.flush();
     }
 }
+
+#[async_trait]
+impl Scratch for NativeStorage {
+    type ScratchStore = TempNativeStore;
+
+    async fn get_scratch_store(&self) -> Result<Self::ScratchStore> {
+        TempNativeStore::new(&self.db)
+    }
+}
+
+#[derive(Clone)]
+/// A [NativeStore] that does not persist data after dropping.
+/// Can be created from [NativeStorage]'s [Scratch] implementation.
+pub struct TempNativeStore {
+    db: Db,
+    name: String,
+    store: NativeStore,
+}
+
+impl TempNativeStore {
+    pub(crate) fn new(db: &Db) -> Result<Self> {
+        let db = db.to_owned();
+        let name = format!("temp-native-store-{}", rand::random::<u32>());
+        let store = NativeStore::new(&db.open_tree(&name)?);
+        Ok(TempNativeStore { db, store, name })
+    }
+}
+
+#[async_trait]
+impl Store for TempNativeStore {
+    async fn read(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        self.store.read(key).await
+    }
+
+    async fn write(&mut self, key: &[u8], bytes: &[u8]) -> Result<Option<Vec<u8>>> {
+        self.store.write(key, bytes).await
+    }
+
+    async fn remove(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        self.store.remove(key).await
+    }
+
+    async fn flush(&self) -> Result<()> {
+        self.store.flush().await
+    }
+}
+
+impl Drop for TempNativeStore {
+    fn drop(&mut self) {
+        if let Err(e) = self.db.drop_tree(&self.name) {
```

Comment:
I wonder if we could better enforce this "storage disappears when dropped" aspect by saying something like:

```rust
pub trait DroppableStore {
  async fn drop_store(&mut self);
}

pub struct TemporaryStore(T) where T: Store + DroppableStore;

impl Store for TemporaryStore { /* ... */ }

impl DroppableStore for TemporaryStore {
  async fn drop_store(&mut self) {
    self.0.drop_store().await;
  }
}

impl Drop for TemporaryStore {
  fn drop(&mut self) {
    self.drop_store();
  }
}
```

And later we can implement async drop for `IndexedDbStore`. Then, we _also_ do a clean up check for droppable stores on every initialization of all storage layers, to cover the possibility that the application exited before async work could be scheduled.

Reply:
I think there's something here, but want to note that the `Drop` trait still needs a blocking executor